### PR TITLE
perf(s.string): speed up URL and IP constraints

### PR DESCRIPTION
**Pull request by kyranet**
Resolving a TODO and then improving IP's error reporting performance

### CODE REVIEW

File: src/constraints/util/urlValidators.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+function allowedProtocolsFn(allowedProtocols: StringProtocol[]) {
+	const oneOf = allowedProtocols.join(', ');
+	return (input: string, url: URL) =>
+		allowedProtocols.includes(url.protocol as StringProtocol)
+			? null
+			: new ConstraintError('s.string.url', 'Invalid URL protocol', input, `expected ${url.protocol} to be one of: ${oneOf}`);
+}
+
+function allowedDomainsFn(allowedDomains: StringDomain[]) {
+	const oneOf = allowedDomains.join(', ');
```
Comment (vladfrangu) [inline src/constraints/util/urlValidators.ts:31]: Can you move this in the error definition instead, as it's a useless join call for valid domains

File: src/constraints/util/urlValidators.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	const fns: ErrorFn<[input: string, url: URL], ConstraintError<string>>[] = [];
+
+	if (options?.allowedProtocols?.length) fns.push(allowedProtocolsFn(options.allowedProtocols));
+	if (options?.allowedDomains?.length) fns.push(allowedDomainsFn(options.allowedDomains));
+
+	return combinedErrorFn(...fns);
+}
+
+function allowedProtocolsFn(allowedProtocols: StringProtocol[]) {
+	const oneOf = allowedProtocols.join(', ');
```
Comment (vladfrangu) [inline src/constraints/util/urlValidators.ts:23]: Ditto

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/constraints/util/urlValidators.ts (hunk 1 above)
Comment (kyranet) [reply src/constraints/util/urlValidators.ts:31]: ... what

Comment (vladfrangu) [reply src/constraints/util/urlValidators.ts:31]: You take this join call
You remove it from here
You move it in the `new Error(` part where its used
You profit

Comment (kyranet) [reply src/constraints/util/urlValidators.ts:31]: Mate, that's not executed in valid domains.

Comment (vladfrangu) [reply src/constraints/util/urlValidators.ts:31]: Useless execution still :)

**Review by vladfrangu — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Pull request merged**